Update the list of supported ClickHouse versions to the latest and relevant metrics

ekopanev opened the pull request:
### What does this PR do?
<!-- A brief description of the change being made with this pull request. -->

Updates the list of supported ClickHouse versions and corresponding metrics - now it can be automatically generated for all the relevant versions and metrics.

### Motivation
<!-- What inspired you to submit this pull request? -->
- currently very old and outdated versions of Clickhouse are supported
- that's why the list of available metrics is very limited
- it was always updated manually, so the list of metrics is also incomplete

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Add the `qa/skip-qa` label if the PR doesn't need to be tested during QA.
- [ ] If you need to backport this PR to another branch, you can add the `backport/<branch-name>` label to the PR and it will automatically open a backport PR once this one is merged

ekopanev committed:
added system_metrics generator and generated queries

ekopanev committed:
updated default hatch env matrix, added keeper repo

ekopanev committed:
update docker compose files, switched from zookeeper to clickhouse-keeper

ekopanev committed:
updated base conftest for new docker compose

ekopanev committed:
added generator for system events

ekopanev committed:
Merge branch 'DataDog:master' into clickhouse_integration

ekopanev committed:
updated generator, added tests template

ekopanev committed:
added generated data

ekopanev committed:
updated metadata.csv

ekopanev committed:
updated tests

ekopanev committed:
added async metrics parsing

ekopanev committed:
added support of plain and json logs by default

ekopanev committed:
fixed linters

ekopanev committed:
exported new dashboard

ekopanev committed:
added 25.8 and removed 25.5 versions metrics

ekopanev committed:
fixed max desc length for metadata file

ekopanev committed:
added metrics based on system.errors

ekopanev committed:
added metrics stat counter

ekopanev committed:
added system errors check to the agent and fixed e2e

ekopanev committed:
explicit query match mapping

ekopanev committed:
added explicit agent check before the next assert

ekopanev committed:
Merge branch 'DataDog:master' into clickhouse_integration

ekopanev committed:
switch system.errors metrics type to monotonic_gauge and extended metadata.csv descriptions with that info

ekopanev committed:
added changelog entry

ekopanev committed:
updated conf.yaml.example file

ekopanev committed:
increased conditions waiting time on the env startup

ekopanev committed:
split compose files for mounted logs testing and the regular run

ekopanev committed:
fixed metrics_metadata

ekopanev committed:
fixed UNKNOWN_IDENTIFIER metrics name

ekopanev committed:
updated conf.example with new ddev

ekopanev committed:
fixed license-headers

ekopanev committed:
small fixes on dashboard panels

ekopanev committed:
Merge branch 'master' into clickhouse_integration

ekopanev committed:
update tests with brand new CH client

ekopanev committed:
Merge branch 'DataDog:master' into clickhouse_integration

ekopanev committed:
linter fix in conftest

ekopanev committed:
removed unfinished test

ekopanev committed:
Merge branch 'master' into clickhouse_integration

dkirov-dd commented on clickhouse/datadog_checks/clickhouse/clickhouse.py:
```diff
@@ -48,9 +48,7 @@ def __init__(self, name, init_config, instances):
                 queries.SystemAsynchronousMetrics,
-                queries.SystemParts,
-                queries.SystemReplicas,
-                queries.SystemDictionaries,
```
request
Please keep backwards compatibility by adding these queries back.

dkirov-dd commented on clickhouse/hatch.toml:
```diff
@@ -1,18 +1,25 @@
 [[envs.default.matrix]]
 python = ["3.13"]
-version = ["18", "19", "20", "21.8", "22.7"]
+version = ["24.8", "25.3", "25.6", "25.7", "25.8"]
```
request
We don't want to break compatibility with older versions, please append a few new versions instead of replacing the old ones.

sangeetashivaji replied:
Releases before 23.8 are currently end-of-life, which might explain these changes — could you confirm if that’s the rationale behind this version update? 
Also, it looks like these are the last three LTS releases (24.8, 25.3, 25.8) and two stable releases (25.7, 25.6). 
Should we be including LTS alone given stable releases supported for 3 months only & so we have fewer intermediate versions?

dkirov-dd commented on clickhouse/scripts/templates/metrics.tpl:
```diff
@@ -0,0 +1,35 @@
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+# This file is autogenerated.
+# To change this file you should edit scripts/templates/metrics.py and then run the following command:
```
```suggestion
# To change this file you should edit scripts/templates/metrics.tpl and then run the following command:
```

dkirov-dd commented on clickhouse/tests/common.py:
```diff
@@ -3,20 +3,31 @@
+COMPOSE_FILE = os.path.join(HERE, 'docker', 'compose.yaml')
+COMPOSE_LOGS_FILE = os.path.join(HERE, 'docker', 'compose-logs.yaml')
+
+CLICKHOUSE_NODE_NUM = 2
```
question
Why did this value get lowered? It used to be 6

dkirov-dd replied:
Nice! 
I was just asking out of curiosity
This seems good!

dkirov-dd reviewed (changes requested):
Thank you for the contribution!
Here are my initial comments.

sangeetashivaji commented:
Thanks @ekopanev for your contributions to this integration.

dkirov-dd reviewed (commented):
I have a new suggestion for the metric names.

ekopanev commented:
hi @dkirov-dd 

> 1. What percentage of these metrics are always available? (in other words, how many of these metrics require a specific condition in order to be emitted by ClickHouse)

Only metrics from `system.metrics` table will exist on 100%, and the rest is optional and depends on certain conditions (eg: perform special query, make an error, perform query in particular CSP, etc)

Currently in `system_metrics.py` we have 431 different metrics (also keep in mind that this is a combination of all the metrics across all supported versions).

According to the generator's script stats we have the following amount of unique metrics per table:
```
➜  clickhouse git:(clickhouse_integration) ✗ hatch run data:generate
The number of metrics:
- async_metrics:  124
- events:  1012
- metrics:  431
- errors:  631
Total: 2198
```

So only ~19.6% are always available.

> 2. How many metrics would a single [instance](https://github.com/DataDog/integrations-core/blob/d8851ac62ae767925735f35ec1c01d9a82d92729/clickhouse/datadog_checks/clickhouse/data/conf.yaml.example#L25) be expected to submit at each check run? For example, is it expected that the clickhouse.asynchronous_metrics.FilesystemCacheBytes will be submitted multiple times with different tags at each run.

As I mentioned above - at least 431 metrics, the rest depends on some particular condition or error.

For example, I've just started a test env
```
ddev env start --dev clickhouse "py3.13-25.8"
```
and run single check right after and got the following number of metrics:
```
> ddev env agent clickhouse py3.13-25.8 check

    clickhouse (5.3.1)
    ------------------
      Instance ID: clickhouse:21251c2b93536980 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/clickhouse.d/clickhouse.yaml[0]
      Total Runs: 1
      Metric Samples: Last Run: 730, Total: 730
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 1
      Average Execution Time : 157ms
      Last Execution Date : 2025-10-31 16:56:48.624 UTC (1761929808624)
      Last Successful Execution Date : 2025-10-31 16:56:48 UTC (1761929808000)
```

In ~10 min it's pretty the same - `Metric Samples: Last Run: 739, Total: 739`

And with some several consecutive queries like:
```
> docker exec -it clickhouse-01 clickhouse-client
...
clickhouse-01 :) select 1;
clickhouse-01 :) select 1;
clickhouse-01 :) select now();
```
It was just a bit more and constant - `Metric Samples: Last Run: 761, Total: 761`

And regarding `clickhouse.asynchronous_metrics.FilesystemCacheBytes` - yes it will be submitted (as well as other system resources related metrics, eg: cpu/memory/etc)
```
> ddev env agent clickhouse py3.13-25.8 check | grep FilesystemCacheBytes
    "metric": "clickhouse.ClickHouseAsyncMetrics_FilesystemCacheBytes",
```

> 3. Would it make sense to "lock" a portion of the newly added metrics behind an instance flag? This can be useful if there are metrics that are more developer-oriented or too niche for most users. I'm looking at metrics like ClickHouseProfileEvents_CoordinatedMergesMergeCoordinatorLockStateForShareMicroseconds.

Do you mean just to add an extra param in the instance config to enable/disable them, so we can manage the list of performing queries in `QueryManager`?

If you think it is necessary we can "lock" something under instance flags.

ekopanev commented on clickhouse/scripts/generate_metrics.py:
```diff
@@ -0,0 +1,552 @@
+PREFIX_CURRENT_METRICS = 'ClickHouseMetrics'
+PREFIX_PROFILE_EVENTS = 'ClickHouseProfileEvents'
+PREFIX_ASYNC_METRICS = 'ClickHouseAsyncMetrics'
+PREFIX_ERRORS = 'ClickHouseErrors'
```
Could you please also recommend me how to properly query the same metric under different names? For example for `MarkCacheFiles`

Currently we have the following mapping ([source code](https://github.com/DataDog/integrations-core/blob/master/clickhouse/datadog_checks/clickhouse/queries.py#L1930C17-L1930C79)):
```
# from system.asynchronous_metrics
'MarkCacheFiles': {'name': 'MarkCacheFiles', 'type': 'gauge'},
```

But I need to have the same metric with a new name (and for sure for many others)
```
# system.asynchronous_metrics
'MarkCacheFiles': {'name': 'ClickHouseAsyncMetrics_MarkCacheFiles', 'type': 'gauge'},
```

Is it possible to achieve in 1 query to ClickHouse and map properly on the agent side? Or there is only one way - to make 2 different queries with different mappings 🤔

ekopanev replied:
> All metrics should still be submitted under their "legacy" names, a deprecation warning should be added to their metric descriptions.

yes, that's what I'm trying to do right now

> Correct me if I'm wrong: if we only keep the new query to ClickHouseAsyncMetrics_MarkCacheFiles then users with old versions of ClickHouse will no longer be able to get that metric.
> 
> If that is the case then we should make both queries.
Maybe some logic can be added to avoid doing both at every check run. Like querying the new metric first and falling back to the old one.

What I can see is that we will do absolutely similar query to fetch that metric:
```
SELECT value, metric FROM system.asynchronous_metrics
```

but intetnal agent's mapping will be different:
- currently we have it under `MarkCacheFiles` name
```
'MarkCacheFiles': {'name': 'MarkCacheFiles', 'type': 'gauge'},
```
- and the new one will be `ClickHouseAsyncMetrics_MarkCacheFiles` (and with our new naming agreement it will be `asynchronous_metrics.MarkCacheFiles`, just in case)
```
'MarkCacheFiles': {'name': 'ClickHouseAsyncMetrics_MarkCacheFiles', 'type': 'gauge'},
```

So yeah - looks like we have to make 2 queries to be able to have 2 different names for 1 metric. And I see pretty similar approach of deprecating some of the metrics, like it was done recently for some typo in Keeper's metric or so.

Maybe we can even deprecate later all "legacy" metrics at all and switch to the new approach (naming + full metric coverage for relevant ClickHouse versions).

dkirov-dd replied:
@ekopanev perfect, that sounds good to me! 👍

ekopanev committed:
Merge branch 'master' into clickhouse_integration

dkirov-dd commented:
Hey @ekopanev,
I'm just checking in to know if you're still interested/working on this feature.

dkirov-dd commented:
@ekopanev good to know!
I responded to your previous comment, apologies for the oversight.

ekopanev committed:
template typo fixes + renamed metrics prefixes and the package name

ekopanev committed:
removed 25.6 and 25.7 versions

ekopanev committed:
simplified docker setup to 1 CH and 1 Keeper nodes

ekopanev committed:
updated docker setup for legacy CH versions

ekopanev committed:
renamed hatch generator script

ekopanev committed:
rolled back old metrics + added version specific ones

ekopanev committed:
fix format

ekopanev committed:
Merge branch 'master' into clickhouse_integration

ekopanev committed:
CI sync

ekopanev committed:
added changelog file

ekopanev committed:
fixed license-headers validator

ekopanev committed:
renamed changelog file

ekopanev committed:
renamed metrics on the dashboard

ekopanev committed:
Merge remote-tracking branch 'origin/master' into clickhouse_integration

ekopanev committed:
fixed some leftovers after merge

ekopanev committed:
updated flush_interval_milliseconds for newer CH versions for DBM integration tests

ekopanev committed:
fixed DBM tests on supported CH versions

ekopanev committed:
fixed formatter

ekopanev committed:
updated license-headers

ekopanev committed:
returned previous manifest metric-check name

ekopanev committed:
removed all quantiles

ekopanev committed:
removed `ddsource` assertion from the tests

ekopanev committed:
return ddsource back to fqt_events as it was before

ekopanev committed:
Merge remote-tracking branch 'origin/master' into clickhouse_integration

ekopanev committed:
added cache for query manager

ekopanev committed:
Merge remote-tracking branch 'origin/master' into clickhouse_integration

ekopanev committed:
Merge remote-tracking branch 'origin/master' into clickhouse_integration

sangeetashivaji commented:
Overall looks good - 
Two open comments that needs code changes - 
1. **system_errors** metric explosion - current implementation creates a separate top-level metric per error code (~1,000 metrics). Team wants it refactored to follow the MySQL pattern: one metric with the error name as a tag, so it's filterable/aggregatable.
 2. Merge` metadata-legacy.csv` into `metadata.csv` — legacy metrics that are still being emitted need to live in the canonical metadata.csv (marked as legacy in the description column) because that file drives Datadog's automated docs and metric summaries.
 
 
 

ekopanev committed:
Merge remote-tracking branch 'origin/master' into clickhouse_integration

ekopanev committed:
added TLS setting after merge conflicts

ekopanev committed:
fixed dbm_integration tests after the merge

ekopanev committed:
added 23.2 and 23.8 to the legacy versions

ekopanev committed:
onboarded new config settings

ekopanev committed:
fixed LogTrace event metric

ekopanev committed:
introduced generic errors.raised metric for all the errors

ekopanev committed:
Merge branch 'master' into clickhouse_integration

ekopanev committed:
set version gating for system errors to 21.3 to have remote column

ekopanev committed:
added test for system errors query, fixed dbm query check

ekopanev committed:
fixed linter

ekopanev commented:
> Overall looks good - Two open comments that needs code changes -
> 
> 1. **system_errors** metric explosion - current implementation creates a separate top-level metric per error code (~1,000 metrics). Team wants it refactored to follow the MySQL pattern: one metric with the error name as a tag, so it's filterable/aggregatable.
>     2. Merge` metadata-legacy.csv` into `metadata.csv` — legacy metrics that are still being emitted need to live in the canonical metadata.csv (marked as legacy in the description column) because that file drives Datadog's automated docs and metric summaries.

1. changed to just 1 metric with tags + added a dedicated test for that (I also changed a bit one check for `test_query_errors_data` test in dmb_integration, because it uses a strict check of errors order, so any failed query in the log before will fail this test - now it should properly check expected error instead of just `[0]`)
2. it was done just for tracking "legacy" metrics, because it's used only for merging into a common one `metadata.csv` (and later we can get rid of it and have/generate only one file)

Also fixed failed e2e on `clickhouse:20` - it was because of our new system-errors query - `remote` column appeared there only starting 21.3.

eric-weaver reviewed (approved):
Thank you for all your efforts working on this and helping us to update support for the Clickhouse integration @ekopanev. This has been a large effort and we appreciate your patience 🙇 👍
